### PR TITLE
Allow use of Auxilliary functions via mavlink and as mission commands

### PR DESCRIPTION
**Pull request by peterbarker**
Similarly to the way we allow them to be used on buttons.

This can allow for enabling and disabling of avoidance during a mission, for example.

Allows functions to be triggered from a GCS without them being on an RC switch.  This might be useful for Sub, perhaps, which doesn't use RC but could benefit from some of the functions.

### CODE REVIEW

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 1]
```diff
@@ -2433,6 +2433,17 @@ void GCS_MAVLINK::send_heartbeat() const
...
 }
 
+MAV_RESULT GCS_MAVLINK::handle_command_do_aux_function(const mavlink_command_long_t &packet)
+{
+    if (packet.param2 > 2) {
+        return MAV_RESULT_DENIED;
+    }
+    const RC_Channel::AUX_FUNC aux_func = (RC_Channel::AUX_FUNC)packet.param1;
+    const RC_Channel::AuxSwitchPos position = (RC_Channel::AuxSwitchPos)packet.param2;
+    rc().do_aux_function(aux_func, position);
```
Comment (tridge) [inline libraries/GCS_MAVLink/GCS_Common.cpp:2443]: this should fail if invalid aux fn

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_Mission/AP_Mission_Commands.cpp [hunk 2]
```diff
@@ -7,6 +7,26 @@
...
+
+    // sanity check the switch position.  Could map from the mavlink
+    // enumeration if we were really keen
+    switch (pos) {
+    case RC_Channel::AuxSwitchPos::HIGH:
+    case RC_Channel::AuxSwitchPos::MIDDLE:
+    case RC_Channel::AuxSwitchPos::LOW:
+        break;
+    default:
+        // warn here?!
```
Comment (rmackay9) [inline libraries/AP_Mission/AP_Mission_Commands.cpp:23]: probably should address this comment..

File: libraries/GCS_MAVLink/GCS_Common.cpp (hunk 1 above)
Comment (peterbarker) [reply libraries/GCS_MAVLink/GCS_Common.cpp]: Done

File: libraries/AP_Mission/AP_Mission_Commands.cpp (hunk 2 above)
Comment (peterbarker) [reply libraries/AP_Mission/AP_Mission_Commands.cpp]: Removed the comment... really not sure it warrants a string.

File: libraries/AP_Mission/AP_Mission.cpp [hunk 3]
```diff
@@ -2243,6 +2256,8 @@ const char *AP_Mission::Mission_Command::type() const
         return "Parachute";
     case MAV_CMD_DO_SPRAYER:
         return "Sprayer";
+    case MAV_CMD_DO_AUX_FUNCTION:
+        return "AuxSwitch";
```
Comment (rmackay9) [inline libraries/AP_Mission/AP_Mission.cpp:2260]: Should this be AuxFunction instead of AuxSwitch?

File: libraries/AP_Mission/AP_Mission.h [hunk 4]
```diff
@@ -148,6 +148,12 @@ class AP_Mission
         uint8_t action;         // action (0 = release, 1 = grab)
     };
 
+    // AUX_FUNCTION command structure
+    struct PACKED AuxFunction {
+        uint16_t function;  // from RC_Channel::AUX_FUNC
+        uint8_t switchpos;     // from RC_Channel::AuxSwitchPos
```
Comment (rmackay9) [inline libraries/AP_Mission/AP_Mission.h:154]: tiny white space issue.

Comment (peterbarker) [reply libraries/AP_Mission/AP_Mission.h]: Fixed.

File: libraries/AP_Mission/AP_Mission.cpp (hunk 3 above)
Comment (peterbarker) [reply libraries/AP_Mission/AP_Mission.cpp]: Fixed.

**Pull request merged**